File: adapters/django/src/fieldseal_django/unindexable.py
"""The `refuse` message for an unindexable value (`docs/12` §10.2).

`encrypt` does not normalize and `blind_index` does, so a value containing a
code point the pinned Unicode version does not define **stores perfectly well
and cannot be fingerprinted** (docs/09 §7.2). Under `on_unindexable="refuse"`
the write is rejected, and §10.2 makes three normative demands of how that
rejection reads. The core's own message meets none of them, and correctly so
-- it is addressed to whoever wired the column:

    value contains U+0378, which is not assigned in Unicode 17.0.0;
    `nfc-casefold-v1` is pinned to that version and cannot index a character
    it does not define

That is the right message for a log and the wrong one for a person typing
their own name into a form. §10.2 requires the message to:

1. **name the character and where it is** -- "somewhere in this field" is not
   something anyone can act on;
2. **put the fault on the system** -- the value is not invalid, it is a name,
   and the pinned tables are behind. Wording that implies a user mistake is
   wrong on the facts;
3. **offer a route that ends with the real value stored** -- which is why
   `refuse` and `bucket` are specified as a pair. `bucket` is the
   operator-applied escape hatch, not a rival philosophy, and "contact
   support" is only honest because an operator can actually move the column.

**Finding the offending character uses the core as the oracle.** The adapter
must not carry its own copy of the Unicode assignment table -- that table
decides index values, so a second copy that drifts is a silent lookup miss
(docs/09 §7.1). Instead each code point is offered to the core's own
normalizer until one is refused. Refusal under `nfc-casefold-v1` is a per
-code-point property (unassigned, or an unpaired surrogate), so probing one
character at a time gives the same answer as probing the whole string.
"""

from __future__ import annotations

from typing import Any

#: Kept short deliberately. A message that scrolls is a message nobody reads,
#: and the actionable part is the character and the route.
_TEMPLATE = (
    "We can't save this {noun} yet. Our systems don't recognise the "
    "character {char} ({ordinal} character). This is a gap on our side, not "
    "a problem with your {noun} -- it's a character we haven't added support "
    "for yet. You can save with a different spelling, or contact support and "
    "we can enter it manually."
)


def locate_unindexable(normalizer: str, value: str) -> tuple[int, str] | None:
    """Return `(index, character)` of the first code point the normalizer
    refuses, or `None` if it refuses none.

    `None` is a real outcome rather than an error path: a normalizer can
    refuse a *string* for a reason no single character carries, and in that
    case the caller falls back to a message that names no position rather
    than inventing one.
    """
    from fieldseal import normalize
    from fieldseal.errors import InvalidArgument

    for i, ch in enumerate(value):
        try:
            normalize(normalizer, ch)
        except InvalidArgument:
            return i, ch
    return None


def _ordinal(n: int) -> str:
    if 10 <= n % 100 <= 20:
        suffix = "th"
    else:
        suffix = {1: "st", 2: "nd", 3: "rd"}.get(n % 10, "th")
    return f"{n}{suffix}"


def user_message(normalizer: str, value: Any, noun: str = "value") -> str:
    """The §10.2-shaped message for a refused value.

    `noun` lets a deployment say "name" where the column holds one; §10.2's
    worked example is a name, and "we can't save this value" is measurably
    colder than "we can't save this name" for the person reading it.
    """
    if not isinstance(value, str):
        # A non-text column cannot reach the Unicode pin at all; if some
        # future normalizer refuses bytes, say so without guessing a position.
        return (
            f"We can't save this {noun} yet -- our systems don't recognise "
            "part of it. This is a gap on our side, not a problem with your "
            f"{noun}. Contact support and we can enter it manually."
        )
    found = locate_unindexable(normalizer, value)
    if found is None:
        return (
            f"We can't save this {noun} yet -- our systems don't recognise "
            "part of it. This is a gap on our side, not a problem with your "
            f"{noun}. Contact support and we can enter it manually."
        )
    index, char = found
    return _TEMPLATE.format(noun=noun, char=char, ordinal=_ordinal(index + 1))


def operator_detail(normalizer: str, value: Any) -> str:
    """The engineer-facing half, for logs and `ValidationError.params`.

    Kept separate from `user_message` on purpose: §10.2's requirement 2 is
    about what the person who typed the value reads, and a code point and a
    Unicode version in the same sentence undo it.
    """
    if not isinstance(value, str):
        return f"normalizer {normalizer!r} refused a non-text value"
    found = locate_unindexable(normalizer, value)
    if found is None:
        return f"normalizer {normalizer!r} refused the value"
    index, char = found
    return (f"normalizer {normalizer!r} refused U+{ord(char):04X} at index "
            f"{index}")


def validation_error(normalizer: str, value: Any, noun: str = "value") -> Any:
    """The `ValidationError` both refusal sites raise.

    Built here rather than at each site because `ValidationError` **drops
    `code` and `params` when its message is a dict** -- so the inner error has
    to carry them, and a caller that builds the dict itself loses the
    machine-readable half without any warning that it did.
    """
    from django.core.exceptions import ValidationError

    return ValidationError(
        user_message(normalizer, value, noun=noun),
        code="fieldseal_unindexable",
        params={"detail": operator_detail(normalizer, value)},
    )

File: adapters/prisma/src/unindexable.ts
/**
 * Unindexable values (`docs/09` §7.2, `docs/13` §9).
 *
 * `encrypt` does not normalize and `blindIndex` does, so a value containing a
 * code point the pinned Unicode version does not define **stores but cannot be
 * fingerprinted**. Under `on_unindexable: "refuse"` (the default) the core
 * throws `INVALID_ARGUMENT` and the extension lets it out of the operation, so
 * the caller sees a rejected write. Under `"bucket"` the core substitutes the
 * column's reserved marker and the write succeeds, and the query path needs no
 * special case -- the same marker is derived for the operand.
 *
 * The important half is the query path, and it is the §10.2 rule this adapter
 * already lives under: this adapter's whole reason for existing is that
 * `prisma-field-encryption` encrypts filter operands and silently returns
 * nothing. Swallowing an unindexable-value error and returning `[]` would be
 * the same failure with a different cause.
 *
 * **The message.** The extension throws; it does not render. But the thrown
 * error MUST carry what a UI needs to build the message in `docs/12` §10.2 --
 * the offending code point and its offset -- because an error that says only
 * "invalid input" forces the application to show that to a person or guess.
 * The three rules: name the character and its position, put the fault on the
 * system, and offer a route that ends with the real value stored.
 *
 * **The gap this used to work around, closed 2026-08-31 (G22, #88).** `docs/09`
 * §7.1 says cores MUST export the assigned-code-point check (`firstUnassigned`
 * / `first_unassigned`) precisely "for adapters that hold the text earlier and
 * can give a better-sited error" -- which is this adapter. Neither core did, so
 * this module parsed the code point and offset out of the core's error
 * *message*: a dependency on prose. Both cores now export it, returning the
 * code point **and its offset in code points**, and this module calls it.
 *
 * Closing it turned up a defect the issue did not know about. The offset regex
 * (`/at index (\d+)/`) matched only the `identity`/bytes path's message.
 * On `nfc-casefold-v1` -- the normalizer every indexed column here declares,
 * and the only one `on_unindexable` governs in practice -- neither core's
 * message carries an offset at all, so `detail.offset` was always `null` and
 * the rendered message named the character without saying where it was.
 * `docs/12` §10.2 requires both. Measured before the fix, not inferred:
 * `{"codePoint":"U+0378","offset":null}`.
 */

import { firstUnassigned, InvalidArgumentError, UNICODE_VERSION } from "@fieldseal/core";

import { FieldsealNotSupported } from "./errors.ts";

/** What a UI needs to render `docs/12` §10.2's message. */
export interface UnindexableDetail {
  /** e.g. `"U+0378"`. */
  readonly codePoint: string | null;
  /** Offset into the value, when the core reported one. */
  readonly offset: number | null;
  /** What to call the value to a person. */
  readonly noun: string;
}

export class FieldsealUnindexable extends FieldsealNotSupported {
  readonly detail: UnindexableDetail;
  constructor(message: string, detail: UnindexableDetail) {
    super(message);
    this.name = "FieldsealUnindexable";
    this.detail = detail;
  }
}

/**
 * Turn the core's refusal into one an application can render.
 *
 * Anything that is not an unindexable-value refusal is returned untouched: a
 * `KEY_UNAVAILABLE` must not be re-dressed as a data-quality problem.
 *
 * **The operand is a parameter because the character and its position come
 * from the core's exported check, not from its error message** (G22, #88).
 * What this function used to do was regex the message for `U+XXXX` and
 * `at index N` -- and the second one never matched on the path that matters.
 * Measured 2026-08-31: under `nfc-casefold-v1`, the only normalizer
 * `on_unindexable` governs in practice, *neither* core's message carries an
 * offset at all (the offset appears only on the `identity`/bytes path, which
 * this column never takes). So the shipped refusal named the character and
 * silently dropped the position, and `docs/12` §10.2 requires both --
 * "somewhere in this field" is not something a person can act on. Exporting
 * the check was the fix; parsing prose could not have been.
 */
export function unindexableError(
  e: unknown,
  label: string,
  noun: string,
  operand: string | Uint8Array,
): unknown {
  if (!(e instanceof InvalidArgumentError)) return e;

  const text = asText(operand);
  const stray = text === null ? undefined : firstUnassigned(text);
  // No offending code point means the refusal was for some other reason --
  // undecodable bytes, an operand the normalizer rejects as a whole. Better
  // discrimination than the old regex, which inferred "unindexable value" from
  // the presence of a `U+` in the text of a message.
  if (stray === undefined) return e;

  const cp = `U+${stray.codePoint.toString(16).toUpperCase().padStart(4, "0")}`;
  const offset = stray.offset;
  const where = ` at position ${String(offset + 1)}`;
  return new FieldsealUnindexable(
    `${label}: this ${noun} contains a character (${cp})${where} that ` +
      `this system cannot index yet, so it cannot be saved on a searchable ` +
      `column. This is a limitation of the system, not a problem with the ` +
      `value: the index is pinned to a published Unicode version, and a ` +
      `character added after that pin has no agreed handling, so two servers ` +
      `would fingerprint it differently and the row would silently stop being ` +
      `findable. The value itself is storable -- an operator can enable ` +
      `\`on_unindexable: "bucket"\` for this column, which keeps the real value ` +
      `and keeps the row findable, at the documented cost in docs/12 §10.4. ` +
      `The index is pinned to Unicode ${UNICODE_VERSION}. Original: ${e.message}`,
    { codePoint: cp, offset, noun },
  );
}

/**
 * The operand as text, or `null` when it is not text at all.
 *
 * A non-string column (`as: "bytes"`, and every numeric rendering) hands the
 * index derivation raw bytes. Decoded strictly: bytes that are not UTF-8 have
 * no characters to count, so "the Nth character" would be an invented number
 * and the caller gets the core's own error instead.
 */
function asText(operand: string | Uint8Array): string | null {
  if (typeof operand === "string") return operand;
  try {
    return new TextDecoder("utf-8", { fatal: true }).decode(operand);
  } catch {
    return null;
  }
}
